### PR TITLE
GO1P-31868: update logic to build flexible link depend on app type

### PR DESCRIPTION
**Pull request by tainbgo1**

### CODE REVIEW

File: tests/portal/PortalCheckerTest.php [hunk 1]
```diff
@@ -470,12 +474,43 @@ public function testSelectedContentSelections(array $collections, int $allowMark
...
 
-        $this->assertEquals('https://qa.go1.cloud/p/#/', (new PortalChecker)->buildLink($portal, '', ''));
+        $this->assertEquals($expect, (new PortalChecker)->buildLink($portal, '', 'p/#'));
+    }
+
+    public function testBuildLinkDefaultPrefix()
+    {
+        putenv("ENV=production");
+        putenv("ENV_HOSTNAME_QA");
+        $instanceId = $this->createPortal($this->go1, ['title' => 'az.mygo1.com']);
```
Comment (drushman) [inline tests/portal/PortalCheckerTest.php]: `$instanceId`  -> `$portalId`

File: tests/portal/PortalCheckerTest.php [hunk 2]
```diff
@@ -470,12 +474,43 @@ public function testSelectedContentSelections(array $collections, int $allowMark
...
+        $portal = PortalHelper::load($this->go1, $instanceId);
+        $this->assertEquals('https://az.mygo1.com/p/#/app/course-overview/123', (new PortalChecker)->buildLink($portal, 'app/course-overview/123'));
+    }
+
+    public function testBuildLinkForMonolith()
+    {
+        putenv("MONOLITH=monolith");
+        putenv("ENV_HOSTNAME=localhost");
+        putenv("ENV_HOSTNAME_QA");
+        $instanceId = $this->createPortal($this->go1, ['title' => 'az.mygo1.com']);
```
Comment (drushman) [inline tests/portal/PortalCheckerTest.php]: Same above

File: tests/portal/PortalCheckerTest.php [hunk 3]
```diff
@@ -470,12 +474,43 @@ public function testSelectedContentSelections(array $collections, int $allowMark
...
+    }
+
+    /**
+     * @dataProvider dataBuildLinkQA
+     */
+    public function testBuildLinkForQA($env, $expect)
+    {
+        putenv("ENV=qa");
+        putenv($env);
+        $instanceId = $this->createPortal($this->go1, ['title' => 'qa.mygo1.com']);
```
Comment (drushman) [inline tests/portal/PortalCheckerTest.php]: Same below

File: portal/PortalChecker.php [hunk 4]
```diff
@@ -129,51 +129,59 @@ public function allowRegister($portal)
...
+        $domain = $this->getDomain($portal, $replacePublicDomain);
+
+        return (PortalHelper::WEBSITE_DOMAIN == $domain)
+            ? "https://{$domain}/{$uri}"
+            : "https://{$domain}/{$prefix}/{$uri}";
+    }
+
+    private function getDomain(stdClass $portal, bool $replacePublicDomain = true): string
+    {
+        $env = (getenv('MONOLITH') && getenv('ENV_HOSTNAME')) ? 'monolith' : (getenv('ENV') ?: 'production');
```
Comment (andytruong) [inline portal/PortalChecker.php]: Is there a method to get env? It's quite messy :/

File: portal/PortalChecker.php [hunk 5]
```diff
@@ -129,51 +129,59 @@ public function allowRegister($portal)
...
-                $domain = PortalHelper::WEBSITE_DEV_INSTANCE . '/p';
+                $domain = PortalHelper::WEBSITE_DEV_INSTANCE;
                 break;
-        }
 
-        if (getenv('MONOLITH') && getenv('ENV_HOSTNAME')) {
-            $domain = getenv('ENV_HOSTNAME') . '/p';
-        }
+            case 'monolith':
+                $domain = getenv('ENV_HOSTNAME');
```
Comment (andytruong) [inline portal/PortalChecker.php]: Too many env variable to manage?

**Pull request merged**